# Transform data

import libraries

In [1]:
from pathlib import Path
import pandas as pd
import numpy as np

set variables

In [2]:
data_dir = Path("../data")
bronze_dir = data_dir / "bronze"
silver_dir = data_dir / "silver"

In [3]:
attack_map = {
    'normal.': 'normal',
    
    # DoS
    'back.': 'dos',
    'land.': 'dos',
    'neptune.': 'dos',
    'pod.': 'dos',
    'smurf.': 'dos',
    'teardrop.': 'dos',
    'mailbomb.': 'dos',
    'processtable.':'dos',
    'udpstorm.':'dos',

    # Probe
    'ipsweep.': 'probe',
    'nmap.': 'probe',
    'portsweep.': 'probe',
    'satan.': 'probe',
    'snmpgetattack.': 'probe',
    'snmpguess.': 'probe',
    'mscan.': 'probe',
    'saint.': 'probe',
    'ps.': 'probe',
    
    # R2L
    'ftp_write.': 'r2l',
    'guess_passwd.': 'r2l',
    'imap.': 'r2l',
    'multihop.': 'r2l',
    'phf.': 'r2l',
    'spy.': 'r2l',
    'warezclient.': 'r2l',
    'warezmaster.': 'r2l',
    'apache2.': 'r2l',
    'httptunnel.': 'r2l',
    'named.': 'r2l',
    'worm.': 'r2l',
    'sqlattack.': 'r2l',
    
    # U2R
    'buffer_overflow.': 'u2r',
    'loadmodule.': 'u2r',
    'perl.': 'u2r',
    'rootkit.': 'u2r',
    'sendmail.': 'u2r',
    'xterm.': 'u2r',
    'xlock.': 'u2r',
    'xsnoop.': 'u2r',
}

In [4]:
columns = [
    "duration",
    "protocol_type",
    "service",
    "flag",
    "src_bytes",
    "dst_bytes",
    "land",
    "wrong_fragment",
    "urgent",
    "hot",
    "num_failed_logins",
    "logged_in",
    "num_compromised",
    "root_shell",
    "su_attempted",
    "num_root",
    "num_file_creations",
    "num_shells",
    "num_access_files",
    "num_outbound_cmds",
    "is_host_login",
    "is_guest_login",
    "count",
    "srv_count",
    "serror_rate",
    "srv_serror_rate",
    "rerror_rate",
    "srv_rerror_rate",
    "same_srv_rate",
    "diff_srv_rate",
    "srv_diff_host_rate",
    "dst_host_count",
    "dst_host_srv_count",
    "dst_host_same_srv_rate",
    "dst_host_diff_srv_rate",
    "dst_host_same_src_port_rate",
    "dst_host_srv_diff_host_rate",
    "dst_host_serror_rate",
    "dst_host_srv_serror_rate",
    "dst_host_rerror_rate",
    "dst_host_srv_rerror_rate", 
    "attack_type"
]

set test data

In [5]:
test_data = pd.read_csv(bronze_dir / "corrected.csv", names=columns)
test_data['attack_category'] = test_data['attack_type'].map(attack_map)
test_data['connection_type'] = np.where(test_data['attack_category'] == 'normal', 'normal', 'attack')
print(f"Corrected data shape: {test_data.shape}")
test_data.head()

Corrected data shape: (311029, 44)


,duration,protocol_type,service,flag,src_bytes,dst_bytes,land,wrong_fragment,urgent,hot,...,dst_host_diff_srv_rate,dst_host_same_src_port_rate,dst_host_srv_diff_host_rate,dst_host_serror_rate,dst_host_srv_serror_rate,dst_host_rerror_rate,dst_host_srv_rerror_rate,attack_type,attack_category,connection_type
0,0,udp,private,SF,105,146,0,0,0,0,...,0.01,0.00,0.0,0.0,0.0,0.0,0.0,normal.,normal,normal
1,0,udp,private,SF,105,146,0,0,0,0,...,0.01,0.00,0.0,0.0,0.0,0.0,0.0,normal.,normal,normal
2,0,udp,private,SF,105,146,0,0,0,0,...,0.01,0.00,0.0,0.0,0.0,0.0,0.0,normal.,normal,normal
3,0,udp,private,SF,105,146,0,0,0,0,...,0.01,0.00,0.0,0.0,0.0,0.0,0.0,snmpgetattack.,probe,attack
4,0,udp,private,SF,105,146,0,0,0,0,...,0.01,0.01,0.0,0.0,0.0,0.0,0.0,snmpgetattack.,probe,attack


Jak widać jest 311029 wierszy i 44 kolumny

print attack by type

In [6]:
print(test_data.groupby(by=['attack_category', 'attack_type']).size())

attack_category  attack_type     
dos              back.                 1098
                 land.                    9
                 mailbomb.             5000
                 neptune.             58001
                 pod.                    87
                 processtable.          759
                 smurf.              164091
                 teardrop.               12
                 udpstorm.                2
normal           normal.              60593
probe            ipsweep.               306
                 mscan.                1053
                 nmap.                   84
                 portsweep.             354
                 ps.                     16
                 saint.                 736
                 satan.                1633
                 snmpgetattack.        7741
                 snmpguess.            2406
r2l              apache2.               794
                 ftp_write.               3
                 guess_passwd.         436

print info about the data

In [7]:
test_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 311029 entries, 0 to 311028
Data columns (total 44 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   duration                     311029 non-null  int64  
 1   protocol_type                311029 non-null  str    
 2   service                      311029 non-null  str    
 3   flag                         311029 non-null  str    
 4   src_bytes                    311029 non-null  int64  
 5   dst_bytes                    311029 non-null  int64  
 6   land                         311029 non-null  int64  
 7   wrong_fragment               311029 non-null  int64  
 8   urgent                       311029 non-null  int64  
 9   hot                          311029 non-null  int64  
 10  num_failed_logins            311029 non-null  int64  
 11  logged_in                    311029 non-null  int64  
 12  num_compromised              311029 non-null  int64  
 13  root_shell

Jak widać wszystkie wartości mają `Non-null` równy do ilości wierszy, więc nie ma wartości nullowych

set train data

In [8]:
train_data = pd.read_csv(bronze_dir / "kddcup_data_10_percent_corrected.csv", names=columns)
train_data['attack_category'] = train_data['attack_type'].map(attack_map)
train_data['connection_type'] = np.where(train_data['attack_category'] == 'normal', 'normal', 'attack')
print(f"Training data shape: {train_data.shape}")
train_data.head()

Training data shape: (494021, 44)


,duration,protocol_type,service,flag,src_bytes,dst_bytes,land,wrong_fragment,urgent,hot,...,dst_host_diff_srv_rate,dst_host_same_src_port_rate,dst_host_srv_diff_host_rate,dst_host_serror_rate,dst_host_srv_serror_rate,dst_host_rerror_rate,dst_host_srv_rerror_rate,attack_type,attack_category,connection_type
0,0,tcp,http,SF,181,5450,0,0,0,0,...,0.0,0.11,0.0,0.0,0.0,0.0,0.0,normal.,normal,normal
1,0,tcp,http,SF,239,486,0,0,0,0,...,0.0,0.05,0.0,0.0,0.0,0.0,0.0,normal.,normal,normal
2,0,tcp,http,SF,235,1337,0,0,0,0,...,0.0,0.03,0.0,0.0,0.0,0.0,0.0,normal.,normal,normal
3,0,tcp,http,SF,219,1337,0,0,0,0,...,0.0,0.03,0.0,0.0,0.0,0.0,0.0,normal.,normal,normal
4,0,tcp,http,SF,217,2032,0,0,0,0,...,0.0,0.02,0.0,0.0,0.0,0.0,0.0,normal.,normal,normal


info about train data

In [9]:
train_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 494021 entries, 0 to 494020
Data columns (total 44 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   duration                     494021 non-null  int64  
 1   protocol_type                494021 non-null  str    
 2   service                      494021 non-null  str    
 3   flag                         494021 non-null  str    
 4   src_bytes                    494021 non-null  int64  
 5   dst_bytes                    494021 non-null  int64  
 6   land                         494021 non-null  int64  
 7   wrong_fragment               494021 non-null  int64  
 8   urgent                       494021 non-null  int64  
 9   hot                          494021 non-null  int64  
 10  num_failed_logins            494021 non-null  int64  
 11  logged_in                    494021 non-null  int64  
 12  num_compromised              494021 non-null  int64  
 13  root_shell

Tutaj też nie ma wartości nullowych

save data to silver directory

In [10]:
silver_dir.mkdir(parents=True, exist_ok=True)

persist_train_data = train_data.drop(columns=['attack_type']).rename(columns={'attack_category':'attack_type'})
persist_test_data = test_data.drop(columns=['attack_type']).rename(columns={'attack_category':'attack_type'})



persist_train_data.to_csv(silver_dir / "train_data.csv", index=False,)
persist_test_data.to_csv(silver_dir / "test_data.csv", index=False)